alternator/streams.cc

```diff
@@ -742,24 +829,51 @@ future<executor::request_return_type> executor::get_records(client_state& client
     auto& db = _proxy.get_db().local();
     schema_ptr schema, base;
     try {
-        auto& log_table = db.find_column_family(iter.shard.table);
+        auto& log_table = db.find_column_family(iter.table);
         schema = log_table.schema();
         base = cdc::get_base_table(db, *schema);
     } catch (...) {        
     }
 
     if (!schema || !base || !is_alternator_keyspace(schema->ks_name())) {
-        throw api_error::resource_not_found(boost::lexical_cast<std::string>(iter.shard.table));
+        throw api_error::resource_not_found(boost::lexical_cast<std::string>(iter.table));
+    }
+
+    auto f = make_ready_future<shard_iterator>(iter);
+    if (!iter.shard.has_expired()) {
+        // ugh. figure out if we have end-of-shard
+        // We try to avoid the extra query by using the
+        // gossip state as a cache check - i.e. know if
+        // topology has changed and invalidated the current
+        // iterator. This is far from exact, so we could still
+        // generate duplicate records at generation changes
+        auto& ep = utils::fb_utilities::get_broadcast_address();
+        auto ts = cdc::get_streams_timestamp_for(ep, _ss.gossiper());
+        if (ts != db_clock::time_point{} &&  ts != iter.shard.start_time) {
+            // this is not really correct, since this should actually mean the
+            // iterator is more than 15 minutes old. But it sort of fits our
+            // use case here, namely the client managed to get and use an iterator
+            // inbetween topology change -> cdc generation change.
```

Comment:
I'm a bit worried about this (ab)use of ExpiredIteratorException... A "normal" application which queries a stream every, say, 10 seconds, does not expect to ever see an ExpiredIteratorException (which suggests the application didn't update the iterator for 15 minutes), so the client code might not be prepared to deal with this case. It would be good if we can find an example stream-using application (I think you have one in Java) and check what it does in this case.

Reply:
At least things like kinesis _has_ to be prepared for it, since its very async nature means it can't really be sure when it is gonna actually perform a task, such as reading from an iterator. 
Again, the above case should be rare, I "expired" seemed the best mapping to the case of an iterator actually going stale. 
The dynamo docs says that the exception is thrown when "The shard iterator has expired and can no longer be used to retrieve stream records. A shard iterator expires 15 minutes after it is retrieved using the GetShardIterator action.". 
I read that as "it will absolutely expire after 15m, but it probably could at other times as well".

Reply:
Ok, I guess we can do this for now. I'm not convinced, though, that your reading of the doc is the correct one. I can also read it as "A shard iterator may expire in 15 minutes so don't think you can hold it for hours" - but an application which thinks it only holds iterators for 10 seconds may not be prepared to handle this event because it never thinks it will run for anything close to 15 minutes. But let's ignore this issue for now.